<launch>
  <!-- change to represent whether stereo camera outputs two images or just one -->
  <arg name="single_image_from_camera" default="true"/>

  <!-- change to a supported pixel format. Check options with v4l2-ctl - -list-formats-ext -->
  <arg name="width" default="1280" />
  <arg name="height" default="480" />
  <arg name="half_width" default="$(eval arg('width') / 2)" />
  <arg name="fps" default="30" />
  <arg name="image_format" default="compressed"/>
  <arg name="format" default="jpeg"/>
  <arg name="skip_frames" default="0"/>

  <!-- <arg name="image_format" default="raw"/>
  <arg name="format" value="YUYV"/> -->

  <!-- Specify device and topic names -->
  <arg name="device_left" default="/dev/video0" />
  <arg name="output_topic_left" default="image_raw" />
  <arg name="device_right" default="/dev/video1" />
  <arg name="output_topic_right" default="image_raw" />

  <!-- arguments for TF publishing -->
  <arg name="frame" default="stereo_camera_frame" />
  <arg name="camera_parent" default="world"/>
  <arg name="transform_yaml" default="package://dvrk_vision/defaults/registration_params_vu.yaml"/>

  <!-- Camera settings are as listed by v4l2-ctl -l -->
  <arg name="exposure" default="150" />
  <arg name="white_balance" default="4000" />
  <arg name="auto_white_balance" default="0" />
  <arg name="auto_exposure" default="1"/>
  <arg name="saturation" default="50" />
  <arg name="brightness" default="0" />
  <arg name="hue" default="-150" />

  <arg name="camera_info_url_left" default="file://$(find dvrk_vision)/defaults/stereo_cam0_calibration_$(arg width)x$(arg height).yaml" />
  <arg name="camera_info_url_right" default="file://$(find dvrk_vision)/defaults/stereo_cam1_calibration_$(arg width)x$(arg height).yaml" />

  <node ns="stereo" pkg="nodelet" type="nodelet" name="standalone_nodelet"  args="manager"/>
  
  <group if="$(arg single_image_from_camera)">

    <node ns="stereo" pkg="nodelet" type="nodelet" name="uvc_camera" args="load uvc_camera/CameraNodelet standalone_nodelet" output="screen">
      <param name="format" value="$(arg format)"/>
      <param name="width" value="$(arg width)" />
      <param name="height" value="$(arg height)" />
      <param name="fps" value="$(arg fps)" />
      <param name="frame_id" value="$(arg frame)" />
      <param name="skip_frames" value="$(arg skip_frames)" />
      <param name="white_balance_temperature" value="$(arg white_balance)"/>
      <param name="auto_white_balance" value="$(arg auto_white_balance)"/>
      <param name="exposure_auto" value="$(arg auto_exposure)"/>
      <param name="exposure_absolute" value="$(arg exposure)"/>
      <param name="saturation" value="$(arg saturation)"/>
      <param name="brightness" value="$(arg brightness)"/>
      <param name="hue" value="$(arg hue)"/>
      <param name="device" value="$(arg device_left)"/>
    </node>

    <node ns="stereo" pkg="nodelet" type="nodelet" args="standalone dvrk_vision/stereo_camera_info" name="cam_info_node">
      <param name="left/camera_info_url" value="$(arg camera_info_url_left)" />
      <param name="right/camera_info_url" value="$(arg camera_info_url_right)" />
    </node>

    <node ns="stereo" pkg="nodelet" type="nodelet" args="standalone image_proc/crop_decimate" name="decimate_l">
      <param name="image_transport" value="$(arg image_format)"/>
      <param name="x_offset" type="int" value="0" />
      <param name="width" type="int" value="$(arg half_width)" />
      <param name="height" type="int" value="$(arg height)" />\
      <remap from="camera/image_raw" to="image_raw"/>
      <remap from="camera/image_info" to="camera_info"/>
      <param name="camera_info_url" type="string" value="$(arg camera_info_url_left)" />
      <remap from="camera_out/image_raw" to="left/image_raw"/>
      <remap from="left/camera_info" to="garbage"/>
    </node>

    <node ns="stereo" pkg="nodelet" type="nodelet" args="standalone image_proc/crop_decimate" name="decimate_r">
      <param name="image_transport" value="$(arg image_format)"/>
      <param name="x_offset" type="int" value="$(arg half_width)" />
      <param name="width" type="int" value="$(arg half_width)" />
      <param name="height" type="int" value="$(arg height)" />
      <param name="camera_info_url" type="string" value="$(arg camera_info_url_right)" />
      <remap from="camera/image_raw" to="image_raw"/>
      <remap from="camera/image_info" to="camera_info"/>
      <remap from="camera_out/image_raw" to="right/image_raw"/>
      <remap from="right/camera_info" to="garbage"/>
    </node>

  </group>

  <group unless="$(arg single_image_from_camera)">

    <node ns="stereo" pkg="nodelet" type="nodelet" name="stereo_cams" args="load uvc_camera/StereoNodelet standalone_nodelet" output="screen">
      <param name="width" value="$(arg width)" />
      <param name="height" value="$(arg height)" />
      <param name="fps" value="$(arg fps)" />
      <param name="frame_id" value="$(arg frame)" />
      <param name="format" value="$(arg format)"/>
      <param name="left/device" value="$(arg device_left)" />
      <param name="right/device" value="$(arg device_right)" />
      <param name="skip_frames" value="$(arg skip_frames)" />
      <param name="left/camera_info_url" value="$(arg camera_info_url_left)" />
      <param name="right/camera_info_url" value="$(arg camera_info_url_right)" />
      <param name="white_balance_temperature" value="$(arg white_balance)"/>
      <param name="auto_white_balance" value="$(arg auto_white_balance)"/>
      <param name="exposure_auto" value="$(arg auto_exposure)"/>
      <param name="exposure_absolute" value="$(arg exposure)"/>
      <param name="saturation" value="$(arg saturation)"/>
      <param name="brightness" value="$(arg brightness)"/>
      <param name="hue" value="$(arg hue)"/>
    </node>

  </group>

  <node ns="stereo" pkg="nodelet" type="nodelet" name="rectify_l"
        args="load image_proc/rectify standalone_nodelet"
        output="screen">
    <param name="image_transport" value="$(arg image_format)"/>
    <param name="queue_size" value="1" />
    <remap from="image_mono" to="left/image_raw"/>
    <remap from="image_rect" to="left/image_rect"/>
  </node>
  <node ns="stereo" pkg="nodelet" type="nodelet" name="rectify_r"
        args="load image_proc/rectify standalone_nodelet"
        output="screen">
    <param name="image_transport" value="$(arg image_format)"/>
    <param name="queue_size" value="1" />
    <remap from="image_mono" to="right/image_raw"/>
    <remap from="image_rect" to="right/image_rect"/>
  </node>
  
  <!-- Camera transform publisher -->
  <node pkg="dvrk_vision" ns="stereo" type="camera_transform_pub.py" name="cam_tf_publisher" output="screen">
    <param name="world_frame" value="$(arg camera_parent)"/>
    <param name="child_frame" value="$(arg frame)"/>
    <param name="transform_yaml" value="$(arg transform_yaml)"/>
  </node>

</launch>
